File: go/utils/file_utils.go
package utils

import (
	"io"
	"os"
	"path/filepath"
	"time"
	"unicode"
)

// The F prefix of a struct name means feature.

type DuplicateFileInfo interface {
	GetSize() int64
	GetPath() string
}

// FDateRangeFileInfo implements DuplicateFileInfo
type FDateRangeFileInfo struct {
	Size         int64
	Path         string
	Name         string
	TimeModified time.Time
}

func (info FDateRangeFileInfo) GetSize() int64 {
	return info.Size
}

func (info FDateRangeFileInfo) GetPath() string {
	return info.Path
}

// FDuplicateFilesFileInfo implements DuplicateFileInfo
type FDuplicateFilesFileInfo struct {
	Size int64
	Path string
}

func (info FDuplicateFilesFileInfo) GetSize() int64 {
	return info.Size
}

func (info FDuplicateFilesFileInfo) GetPath() string {
	return info.Path
}

type FTextFilesFileInfo struct {
	Name         string
	AbsolutePath string
}

type CompleteFileInfo struct {
	Name                  string
	AbsoluteDirectoryPath string
	AbsolutePath          string
	TimeModified          time.Time
	Size                  int64
	IsDirectory           bool
}

type FileData struct {
	Content          string
	CompleteFileInfo CompleteFileInfo
}

type FileSystemNode struct {
	Path        string
	IsDirectory bool
}

type (
	fileFilterMode int
	fileType       int
)

const (
	files fileFilterMode = iota
	NonZeroByteFiles
	FilesAndDirectories
	NonZeroByteFilesAndDirectories
	Directories
)

const (
	AllFiles fileType = iota
	TextFiles
)

const FilePathSeparator = string(filepath.Separator)

func CreateDirectory(filePath string) error {
	return os.Mkdir(filePath, 0755)
}

func IsTextFile(filePath string) (bool, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return false, err
	}
	defer file.Close()

	// Read the first 512 or less to check for non-text characters.
	// DetectContentType of package 'net/http' works with a similar check.
	bytes := make([]byte, 512)
	numberOfBytesRead, err := file.Read(bytes)
	if err != nil && err != io.EOF {
		return false, err
	}
	for _, byte := range bytes[:numberOfBytesRead] {
		if !unicode.IsPrint(rune(byte)) && !unicode.IsSpace(rune(byte)) {
			return false, nil
		}
	}
	return true, nil
}

func resolveDirectoryPath(filePath string, isDirectory bool) string {
	directoryPath := filePath

	if !isDirectory {
		directoryPath = filepath.Dir(filePath)
	}

	return directoryPath
}

func FilterAndHandleFileInfo(
	info os.FileInfo, mode fileFilterMode, fileType fileType, absoluteFilePath string, handler func(CompleteFileInfo),
) error {
	isDir, isRegularFile := info.IsDir(), info.Mode().IsRegular()

	var size int64
	if isRegularFile {
		size = info.Size()
	}

	// is file check
	if isRegularFile && mode == Directories {
		return nil
	}

	// is directory check
	if isDir && (mode == files || mode == NonZeroByteFiles) {
		return nil
	}

	// is zero byte file check
	if isRegularFile && size == 0 && (mode == NonZeroByteFiles || mode == NonZeroByteFilesAndDirectories) {
		return nil
	}

	// is text file check
	if fileType == TextFiles {
		isTextFile, err := IsTextFile(absoluteFilePath)
		if err != nil || !isTextFile {
			return err
		}
	}

	handler(CompleteFileInfo{
		Name:                  info.Name(),
		AbsoluteDirectoryPath: resolveDirectoryPath(absoluteFilePath, isDir),
		AbsolutePath:          absoluteFilePath,
		TimeModified:          info.ModTime(),
		Size:                  size,
		IsDirectory:           isDir,
	})

	return nil
}

func WalkFilterAndHandleFileInfoDirectory(
	filePath string, mode fileFilterMode, fileType fileType, handler func(CompleteFileInfo)) error {
	return filepath.Walk(filePath, func(absoluteFilePath string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		return FilterAndHandleFileInfo(info, mode, fileType, absoluteFilePath, handler)
	})
}

func WalkFilterAndHandleFileInfo(
	node FileSystemNode, mode fileFilterMode, fileType fileType, handler func(CompleteFileInfo),
) error {
	if node.IsDirectory {
		return WalkFilterAndHandleFileInfoDirectory(node.Path, mode, fileType, handler)
	} else {
		info, err := os.Stat(node.Path)
		if err != nil {
			return err
		}
		return FilterAndHandleFileInfo(info, mode, fileType, node.Path, handler)
	}
}

func FileExists(filePath string) (bool, error) {
	_, err := os.Stat(filePath)
	if err == nil {
		return true, nil
	} else if os.IsNotExist(err) {
		return false, nil
	}
	return false, err
}

File: kotlin/src/main/kotlin/utils/FileUtils.kt
package org.example.utils

import java.io.File
import java.net.URLConnection
import java.nio.file.Path
import java.nio.file.attribute.FileTime
import java.time.Instant
import kotlin.io.path.getLastModifiedTime

// The F prefix of a class name means feature.

interface DuplicateFileInfo {
  val file: File
  val size: Long
}

data class FDuplicateFilesFileInfo(
  override val file: File,
  override val size: Long,
) : DuplicateFileInfo

data class FDateRangeFileInfo(
  override val file: File,
  override val size: Long,
  val path: Path,
  val timeModified: Instant,
  var newName: String? // We need the 'newName' property because we cannot change the name of a File instance directly.
) : DuplicateFileInfo

data class FTextFilesFileInfo(
  val file: File,
  val absolutePath: Path
)

data class CompleteFileInfo(
  val file: File,
  val size: Long,
  var absolutePath: Path,
  var absoluteDirectoryPath: Path,
  val timeModified: FileTime?,
)

data class FileData(
  val content: String,
  val completeFileInfo: CompleteFileInfo
)

enum class FileFilterMode {
  FILES,
  NON_ZERO_BYTE_FILES,
  FILES_AND_DIRECTORIES,
  NON_ZERO_BYTE_FILES_AND_DIRECTORIES,
  DIRECTORIES
}

enum class FileType {
  ALL_FILES,
  TEXT_FILES
}

fun isTextFile(file: File): Result<Boolean> = runCatching {
  val mimeType = file.inputStream().use { inputStream ->
    URLConnection.guessContentTypeFromStream(inputStream)
  }

  mimeType?.startsWith("text") == true
}

fun resolveDirectoryPath(filePath: Path, isDirectory: Boolean): Path {
  return if (isDirectory || filePath.parent == null) {
    filePath
  } else {
    filePath.parent
  }
}

fun filterAndHandleFileInfo(
  file: File, mode: FileFilterMode, type: FileType, absoluteFilePath: Path, handler: (CompleteFileInfo) -> Unit
): Result<Unit> = runCatching {
  val size = if (file.isFile) file.length() else 0L

  // is file check
  if (file.isFile && mode == FileFilterMode.DIRECTORIES) {
    return@runCatching
  }

  // is directory check
  if (file.isDirectory && (mode == FileFilterMode.FILES || mode == FileFilterMode.NON_ZERO_BYTE_FILES)) {
    return@runCatching
  }

  // is zero byte file check
  if (file.isFile && size == 0L &&
    (mode == FileFilterMode.NON_ZERO_BYTE_FILES || mode == FileFilterMode.NON_ZERO_BYTE_FILES_AND_DIRECTORIES)) {
    return@runCatching
  }

  // is text file check
  if (type == FileType.TEXT_FILES) {
    if (!isTextFile(file).getOrThrow()) {
      return@runCatching
    }
  }

  handler(CompleteFileInfo(
    file = file,
    size = file.length(),
    absolutePath = absoluteFilePath,
    absoluteDirectoryPath = resolveDirectoryPath(absoluteFilePath, file.isDirectory),
    timeModified = absoluteFilePath.getLastModifiedTime()
  ))
}

fun walkFilterAndHandleFileInfo(
  absoluteFilePath: Path,
  mode: FileFilterMode,
  type: FileType,
  handler: (CompleteFileInfo) -> Unit
): Result<Unit> = runCatching {
  val rootFile = absoluteFilePath.toFile()

  if (!rootFile.isFile && !rootFile.isDirectory) {
    return@runCatching
  }

  val files = if (rootFile.isDirectory) rootFile.walk() else sequenceOf(rootFile)
  files.forEach { file ->
    filterAndHandleFileInfo(file, mode, type, absoluteFilePath, handler).onFailure { throw it }
  }
}
